### ISSUE TITLE
Question: I have a potentially long running operation as part of my extension's installation. Where is the best place to do it?

### ISSUE THREAD

**Issue by balajikris** [Microsoft/vscode#15670]
I have a long running operation (searching if a certain npm-package is globally installed, if not, installing it globally) that I need to do as a part of installing my VSCode extension. 

I'm seeking information about what's the best place to do it? 

Initially, I had it in extension activation (with checks to bail early if package is present) and it worked okay. The flipside is that the very first time the user installs the extension, it won't work until the long running operation completes.

So, I'm moving it as a script to be executed during "preinstall" or "postinstall" step. 
```
    "scripts": {
        "preinstall": "node ./install.js"
    },
```

Now, when I Install my extension (from a packaged vsix), I am observing that the npm-package isn't being installed. So, I'm curious if "scripts:{ }" is the right way to go about this? [To be sure, the same script succeeds, if invoked separately from a command prompt.]

I can provide the script i'm using, if required.

Thanks.

**Comment by balajikris** [comment]
Ah.. I just realized that "preinstall" is picked up by npm and is executed when i do an "npm install" before packing a vsix. But that would install the dependency on my machine. Is there a way for me to direct VSCode to do that on the user's box while installing the extension? I'm only seeing a "Vscode.prepublish" which I don't think would help my case.

So, I wonder what's the right way to go about this?

**Comment by ramya-rao-a** [comment]
Doesn't adding to `dependencies` in your `package.json` help?

**Comment by balajikris** [comment]
In my case, I would like the dependency to be installed in the global location. Is that what it does? I vaguely recall trying that unsuccessfully..

**Comment by balajikris** [comment]
just verified: adding it to `dependencies` brings it to the `node_modules` folder of my extension. I need it to be installed in npm's global `node_modules` folder.

**Comment by joaomoreno** [comment]
Our current recommendation is to do it in activation, like you had before. One suggestion is to indicate to the user that the extension needs to perform some one-time startup tasks, via the status bar or notification messages.

We don't support installation-time tasks.

---

At the same time... why do you need the dependency specifically installed globally?

**Comment by balajikris** [comment]
I'm building a starter kit for something, which along some features like "project scaffolding, snippets, some completionlist help" etc. The project scaffolding is on top of `yeoman` and is effectively a npm package. 

Installing my extension, should install that package to the global folder because:

1. `vscode-yo` looks for generators in the global npm location. 
2. the generator should also be callable from command line with `yo` syntax, which also needs the package to be installed globally.

effectively, the issue is, the package my extension brings along isn't really used by my extension, but just brings in some stuff which may be used independently, later on, outside of my extension, either from VSCode itself or from command line.

for the first time though, this could actually take several minutes, to bring down dependencies via the network. So, doing this operation on activation doesn't result in a great end user experience. While I'll go with that route for now, I think this makes a case for adding support for "installation-time tasks". I'll open a new feature-request it seems agreeable.

/cc: @waderyan

**Comment by joaomoreno** [comment]
What will happen to users that don't have Node installed and install your extension?

It seems to me your extension should describe some Prerequisites it its README, so users know what to install beforehand. Additionally, in a situation that Node exists but the global modules are not installed, it should offer the user the optional service of installing them.

**Comment by balajikris** [comment]
> What will happen to users that don't have Node installed and install your extension?

isn't this all the more reason to provide install time hooks, so that extension writers can check prerequisites and bail early if prerequisites are not satisfied?

> in a situation that Node exists but the global modules are not installed, it should offer the user the optional service of installing them.

Could you please expand a bit more on what is the recommended way to accomplish this? this is what I'm trying to accomplish and thought it would be nice to do so at the time the extension is being installed?

Also, please let me know if I should open a new issue, as we seem to be discussing on a thread that's marked as "closed".

**Comment by joaomoreno** [comment]
Let me rephrase: If users don't have Node installed, your extension couldn't install the global modules it needs at all.

For users that have Node installed: your extension, upon detecting that the global modules it needs are not available, when asked to perform any functionality (running a command, for example), it can propose to the user "Hey do you want to install this and that globally? Press Yes", proceed to install the dependencies and then performs the functionality it was asked to perform.
